File: v2/114.py





# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        # 6 star, 没理解，需多看
        # 先序遍历的非递归实现，prev记录先序遍历的上一个，cur是当前的，把prev的左子树指向None，右子树指向cur即可，然后prev=cur
        if not root:
            return
        stack = [root]
        prev = TreeNode(None)
        while stack:
            cur = stack.pop()
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
            prev.left, prev.right = None, cur
            prev = cur

File: go/114.go
//package main
//
//import "fmt"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

func flatten(root *TreeNode)  {
	if root == nil {return }
	var cur *TreeNode
	stack := []*TreeNode {root}
	prev := &TreeNode{0, nil, nil}
	for len(stack) > 0 {
		cur, stack = stack[len(stack)-1], stack[:len(stack)-1]
		if cur.Right != nil {
			stack = append(stack, cur.Right)
		}
		if cur.Left != nil {
			stack = append(stack, cur.Left)
		}
		prev.Left, prev.Right = nil, cur
		prev = cur

	}

}


//func main() {
//	a := TreeNode{1, nil, nil}
//	b := TreeNode{2, nil, nil}
//	c := TreeNode{3, nil, nil}
//	a.Right = &b
//	a.Left = &c
//	flatten(&a)
//	fmt.Println(a)
//}
//
